code/modules/mob/living/carbon/carbon.dm

```diff
@@ -705,28 +708,50 @@
 	if(unslippable) //if unslippable, don't even bother making checks
 		return FALSE
 
+	var/slide_dir = dir
 	switch(P.wet)
 		if(TURF_WET_WATER)
 			if(Slip(stun_amount = 5, weaken_amount = 3, slip_on_walking = FALSE, overlay_type = TURF_WET_WATER, onwhat = "the wet floor", otherscansee = TRUE, spanclass = "warning"))
-				step(src, dir)
+				step(src, slide_dir)
 			else
 				return FALSE
 
 		if(TURF_WET_LUBE)
-			step(src, dir)
+			step(src, slide_dir)
 			if(Slip(stun_amount = 5, weaken_amount = 3, slip_on_walking = TRUE, overlay_type = TURF_WET_LUBE, slip_on_magbooties = TRUE, onwhat = "the floor", otherscansee = TRUE, spanclass = "warning"))
 				for(var/i = 1 to 4)
 					spawn(i)
 						if(!locked_to)
-							step(src, dir)
+							step(src, slide_dir)
 				take_organ_damage(2) // Was 5 -- TLE
 			else
 				return FALSE
 
 
 		if(TURF_WET_ICE)
-			if(prob(30) && Slip(stun_amount = 4, weaken_amount = 3,  overlay_type = TURF_WET_ICE, onwhat = "the icy floor", otherscansee = TRUE, spanclass = "warning"))
-				step(src, dir)
+			if(!ice_sliding && prob(15) && Slip(stun_amount = 4, weaken_amount = 3,  overlay_type = TURF_WET_ICE, onwhat = "the icy floor", otherscansee = TRUE, spanclass = "warning"))
+				ice_sliding = 1
+				// Wait movement_delay(), which is how long it takes for the movement step onto the slippery tile to finish.
+				spawn(ceil(movement_delay()))
+					var/slide_speed = movement_delay()*4
+					var/turf/next_turf = get_turf(src)
+					// Loop until you reach a tile that doesn't have ice on it. Speed progressively gets slower. next_turf represents the turf you expect to be on
+					// after the next step completes - if you're not on that, it indicates the user manually moved and thus the slide is over.
+					while(can_apply_inertia())
+						if(next_turf != get_turf(src))
+							break
+						next_turf = get_step(src, slide_dir)
+						set_glide_size(DELAY2GLIDESIZE(slide_speed))
+						step(src, slide_dir)
+						sleep(ceil(slide_speed))
+						slide_speed *= 1.1
+						var/list/next_puddles = next_turf.search_contents_for(/obj/effect/overlay/puddle/ice)
+						if( isemptylist(next_puddles) )
+							break
+						var/obj/effect/overlay/puddle/ice/next = next_puddles[1];
```

Comment:
```suggestion
						var/obj/effect/overlay/puddle/ice/next = locate(/obj/effect/overlay/puddle/ice) in next_turf
						if(!next)
							break
```
`search_contents_for()` is recursive, meant for things like checking traitor steal objectives